pkg/cloud/services/instancestate/rule.go

```diff
@@ -0,0 +1,214 @@
+	var ruleNotFound bool
+	ruleResp, err := s.EventBridgeClient.DescribeRule(&eventbridge.DescribeRuleInput{
+		Name: aws.String(s.getEC2RuleName()),
+	})
+	if err != nil {
+		if resourceNotFoundError(err) {
+			ruleNotFound = true
+		} else {
+			return errors.Wrapf(err, "unable to describe rule %s", s.getEC2RuleName())
+		}
+	}
+
+	if ruleNotFound {
+		err = s.createRule()
+		if err != nil {
+			return errors.Wrap(err, "unable to create rule")
+		}
+		// fetch newly created rule
+		ruleResp, err = s.EventBridgeClient.DescribeRule(&eventbridge.DescribeRuleInput{
+			Name: aws.String(s.getEC2RuleName()),
+		})
+
+		if err != nil {
+			return errors.Wrapf(err, "unable to describe new rule %s", s.getEC2RuleName())
+		}
+	}
+
+	queueURLResp, err := s.SQSClient.GetQueueUrl(&sqs.GetQueueUrlInput{
+		QueueName: aws.String(GenerateQueueName(s.scope.Name())),
+	})
+
+	if err != nil {
+		return errors.Wrap(err, "unable to get queue URL")
+	}
+	queueAttrs, err := s.SQSClient.GetQueueAttributes(&sqs.GetQueueAttributesInput{
+		AttributeNames: aws.StringSlice([]string{sqs.QueueAttributeNameQueueArn, sqs.QueueAttributeNamePolicy}),
+		QueueUrl:       queueURLResp.QueueUrl,
+	})
+
+	if err != nil {
+		return errors.Wrap(err, "unable to get queue attributes")
+	}
+
+	targetsResp, err := s.EventBridgeClient.ListTargetsByRule(&eventbridge.ListTargetsByRuleInput{
+		Rule: aws.String(s.getEC2RuleName()),
+	})
+	if err != nil {
+		return errors.Wrapf(err, "unable to list targets for rule %s", s.getEC2RuleName())
+	}
+
+	targetFound := false
+	for _, target := range targetsResp.Targets {
+		// check if queue is already added as a target
+		if *target.Id == GenerateQueueName(s.scope.Name()) && *target.Arn == *queueAttrs.Attributes[sqs.QueueAttributeNameQueueArn] {
+			targetFound = true
+		}
+	}
+
+	if !targetFound {
+		_, err = s.EventBridgeClient.PutTargets(&eventbridge.PutTargetsInput{
+			Rule: ruleResp.Name,
+			Targets: []*eventbridge.Target{{
+				Arn: queueAttrs.Attributes[sqs.QueueAttributeNameQueueArn],
+				Id:  aws.String(GenerateQueueName(s.scope.Name())),
+			}},
+		})
+
+		if err != nil {
+			return errors.Wrapf(err, "unable to add SQS target %s to rule %s", GenerateQueueName(s.scope.Name()), s.getEC2RuleName())
+		}
+	}
+
+	if queueAttrs.Attributes[sqs.QueueAttributeNamePolicy] == nil {
+		// add a policy for the rule so the rule is authorized to emit messages to the queue
+		err = s.createPolicyForRule(&createPolicyForRuleInput{
+			QueueArn: *queueAttrs.Attributes[sqs.QueueAttributeNameQueueArn],
+			QueueURL: *queueURLResp.QueueUrl,
+			RuleArn:  *ruleResp.Arn,
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s Service) createRule() error {
+	eventPattern := eventPattern{
+		Source:     []string{"aws.ec2"},
+		DetailType: []string{Ec2StateChangeNotification},
+		EventDetail: &eventDetail{
+			States: []infrav1.InstanceState{infrav1.InstanceStateShuttingDown, infrav1.InstanceStateTerminated},
```

Comment:
Since we get all messages at once at each interval (tick), there is no api limit issue if we watch all events, right? It'd be good to see the exact state of ec2 instances. This can be tracked in a follow up issue.

Reply:
yeah I don't think there's rate limiting for the sqs api. I know ec2 does rate limiting. With SQS there's a different non-free tier after a certain number of requests. Seeing the ec2 state would definitely be useful.